### ISSUE TITLE
Deserialization in Unity using Converters

### ISSUE THREAD

**Issue by zory** [jilleJr/Newtonsoft.Json-for-Unity.Converters#43]
Hello,
Probably I'm stuck in obvious case, because I can't find any documentation on this.. Maybe you could provide some example?
I'm trying to serialize/deserialize objects which contains List<Vector3Int>

I'm getting
`
ArgumentException: Could not cast or convert from System.String to UnityEngine.Vector3Int.
`

My code:
`
private static JsonSerializerSettings settings = new JsonSerializerSettings()
{
  ContractResolver = new Newtonsoft.Json.UnityConverters.UnityTypeContractResolver(),
  Converters = new JsonConverter[] {
    new Newtonsoft.Json.UnityConverters.Math.Vector3IntConverter(),
  }
};
//--- and the serialization/deserialization
var strVal = JsonConvert.SerializeObject(instance, settings);
var instance = JsonConvert.DeserializeObject<T>(strVal, settings);
`

Thanks in advance.

**Comment by jilleJr** [comment]
Hi @zory! Thanks for reporting this

Though I'm having troubles understanding the full picture. Will try to reproduce it myself and hopefully find some issue.

Could you post the full stacktrace of the error?

**Comment by jilleJr** [comment]
Well something I can tell you right now is that the `UnityTypeContractResolver` type does currently not handle `ISerializationCallbackReceiver`s. If you're depending on that then that's probably the reason this doesn't work.

It can be a future feature, but it's not available right now

**Comment by zory** [comment]
I tried removing everything, leaving just simple types, so it doesn't rely on `ISerializationCallbackReceiver` anymore.
`Dictionary<Vector3Int, int>` -> still Fails. But `Vector3Int` or `List<Vector3Int>` is working as expected.
Maybe I need somehow explicitly state to use some DictionaryConverter or something?

**Comment by zory** [comment]
Yes, your workaround works. Serializing for example List<KeyValuePair<Vector3Int, int>> yields expected results. Thank you for letting me use your brain : ). If you would come up with another solution, let me know : ).

**Comment by MrFJ** [comment]
Hey, I just ran into a similar issue, though your solution does not seem to work for me. This:
`public IEnumerable<KeyValuePair<Vector3Int, string>> TilesSerialized { get; set; } = new Dictionary<Vector3Int, string>();`
is serialized into this:
`"EntitiesSerialized": {
    "(4, 14, 0)": "PassiveEnemyShip",
    "(3, 11, 0)": "PassiveEnemyShip",
    "(1, 9, 0)": "PassiveEnemyShip"
  }`
Which JsonConvert refuses to deserialize. Any thoughts? I even tried adding the Vector3IntConverter manually to the serialize and deserialize calls.

**Comment by jilleJr** [comment]
Hi @MrFJ ! You're using a complex type as a dictionary key. That doesn't work in JSON, and Newtonsoft.JSON's solution for when doing this seems to just be to `.ToString()` the value. Parsing that is not an option.

You do save the type as `IEnumerable<KeyValuePair<Vector3Int, string>>`, but the underlying type that Newtonsoft.JSON sees is still the `Dictionary<,>`

Maybe you could try use an explicit list instead?

```cs
public List<KeyValuePair<Vector3Int, string>> EntitiesSerialized { get; set; } = new Dictionary<Vector3Int, string>().ToList();
```

**Comment by MrFJ** [comment]
Thanks for the quick response! I changed my implementation to: 
`public Dictionary<Vector3Int, string> Entities { get; set; } = new ();
 public IEnumerable<KeyValuePair<Vector3Int, string>> EntitiesSerialized { get => Entities; set => Entities = (Dictionary<Vector3Int, string>)value; }` to no avail, but it turns out ToString() was indeed the missing piece!
 This works perfectly:
 `[JsonIgnore]
 public Dictionary<Vector3Int, string> Entities { get; set; } = new ();

public IEnumerable<KeyValuePair<Vector3Int, string>> EntitiesSerialized { get => Entities.ToList(); set => Entities = (Dictionary<Vector3Int, string>)value; }`

**Comment by jilleJr** [comment]
Great! Happy you found a solution :)

**Comment by MrFJ** [comment]
It seems I spoke too soon. This is my JSON: `"EntitiesSerialized": [
    {
      "Key": {
        "x": 6,
        "y": 12,
        "z": 0
      },
      "Value": "PassiveEnemyShip"
    },
    {
      "Key": {
        "x": 6,
        "y": 11,
        "z": 0
      },
      "Value": "PassiveEnemyShip"
    },
    {
      "Key": {
        "x": 6,
        "y": 8,
        "z": 0
      },
      "Value": "PassiveEnemyShip"
    }
  ]`
  The dictionary is empty after serialization though :/ I even tried the following to no results: `private IEnumerable<KeyValuePair<Vector3Int, string>> entities;
        public IEnumerable<KeyValuePair<Vector3Int, string>> EntitiesSerialized { get => Entities.ToList(); set => entities = value; }`
I'm getting nada :(

**Comment by jilleJr** [comment]
It's getting complicated with the types here. To simplify, you could have it more explicit. Such as by using one class for serializing/deserializing, and one that you're working with otherwise.

Such pattern is usually called a DTO (Data-Transfer-Object).

So that before you serialize it, you always convert the dictionary into a `List<KeyValuePair<,>>`, and then after deserializing you always convert it back into a dictionary. You have then two classes, one that you use during runtime (such as a ScriptableObject or GameObject), and then as soon as you're serializing/deserializing you create this DTO and copy over the data to/from it.

Example:

https://dotnetfiddle.net/W4gUHS
